[AOTI] Support multi-arch when using package_cpp_only

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #154414

Summary: Add support of multi_arch_kernel_binary in the package_cpp_only mode. More specifically, generate specific cmake targets to compile .ptx to .fatbin and embed them in the final shared library or binary.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

desertfire committed:
[AOTI] Support multi-arch when using package_cpp_only

Summary: Add support of multi_arch_kernel_binary in the package_cpp_only mode. More specifically, generate specific cmake targets to compile .ptx to .fatbin and embed them in the final shared library or binary.

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Support multi-arch when using package_cpp_only"


Summary: Add support of multi_arch_kernel_binary in the package_cpp_only mode. More specifically, generate specific cmake targets to compile .ptx to .fatbin and embed them in the final shared library or binary.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D75452096).

desertfire committed:
Update on "[AOTI] Support multi-arch when using package_cpp_only"


Summary: Add support of multi_arch_kernel_binary in the package_cpp_only mode. More specifically, generate specific cmake targets to compile .ptx to .fatbin and embed them in the final shared library or binary.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D75452096](https://our.internmc.facebook.com/intern/diff/D75452096)

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge

desertfire commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but it is failing ROCm trunk job' -c weird

inductor/test_aot_inductor_package.py::TestAOTInductorPackageCpp_cuda::test_compile_after_package [GH job link](https://github.com/pytorch/pytorch/actions/runs/15289254655/job/43006612419) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/a84d8c4a1cc515db274366537afd0b1492800c2d)

The infra error from ROCm probably hide the real failure

desertfire committed:
Update on "[AOTI] Support multi-arch when using package_cpp_only"


Summary: Add support of multi_arch_kernel_binary in the package_cpp_only mode. More specifically, generate specific cmake targets to compile .ptx to .fatbin and embed them in the final shared library or binary.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D75452096](https://our.internmc.facebook.com/intern/diff/D75452096)

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Support multi-arch when using package_cpp_only"


Summary: Add support of multi_arch_kernel_binary in the package_cpp_only mode. More specifically, generate specific cmake targets to compile .ptx to .fatbin and embed them in the final shared library or binary.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

Differential Revision: [D75452096](https://our.internmc.facebook.com/intern/diff/D75452096)

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge
